compiler/rustc_ty_utils/src/layout_sanity_check.rs

```diff
@@ -296,4 +297,50 @@ pub(super) fn sanity_check_layout<'tcx>(
             }
         }
     }
+
+    // If the struct tail is sized and can be unsized, check that unsizing doesn't move the fields around.
+    if let ty::Adt(def, substs) = *layout.ty.kind()
+        && def.is_struct()
+        && let Some(tail_def) = def.variant(FIRST_VARIANT).fields.raw.last()
+        && let tail_binder = cx.tcx.type_of(tail_def.did)
+        && tail_binder.subst(cx.tcx, substs).is_sized(cx.tcx, cx.param_env)
+        && !tail_binder.subst_identity().is_sized(cx.tcx, cx.tcx.param_env(def.did()))
+        && let ty::Param(tail_param) = tail_binder.subst_identity().kind()
+        // `Sync` is chosen arbitrarily here, this could be any trait.
```

Comment:
Could this run into problems with negative trait impls?

Reply:
Looks like the current implementation can break with specialization:

```rust
#![feature(specialization, offset_of, rustc_attrs)]

trait Tr {
    type Assoc;
}

impl<T: ?Sized> Tr for T {
    default type Assoc = ();
}

impl Tr for *mut () {
    type Assoc = i32;
}

struct S<T: ?Sized> {
    x: <T as Tr>::Assoc,
    y: i32,
    tail: T,
}

fn main() {
    dbg!(core::mem::offset_of!(S<*mut ()>, y));
    dbg!(core::mem::offset_of!(S<dyn Sync>, y));
}
```

```text
error: internal compiler error: compiler/rustc_ty_utils/src/layout_sanity_check.rs:336:17: unsizing S<*mut ()> to S<dyn std::marker::Sync> changed field order!
                                TyAndLayout { ty: S<*mut ()>, layout: Layout { size: Size(16 bytes), align: AbiAndPrefAlign { abi: Align(8 bytes), pref: Align(8 bytes) }, abi: Aggregate { sized: true }, fields: Arbitrary { offsets: [Size(0 bytes), Size(4 bytes), Size(8 bytes)], memory_index: [0, 1, 2] }, largest_niche: None, variants: Single { index: 0 } } }
                                TyAndLayout { ty: S<dyn std::marker::Sync>, layout: Layout { size: Size(4 bytes), align: AbiAndPrefAlign { abi: Align(4 bytes), pref: Align(8 bytes) }, abi: Aggregate { sized: false }, fields: Arbitrary { offsets: [Size(0 bytes), Size(0 bytes), Size(4 bytes)], memory_index: [0, 1, 2] }, largest_niche: None, variants: Single { index: 0 } } }
```

This needs a better check for whether the struct can be unsized than just looking at the tail, that would also fix the CI failure.